src/runner/worker.rs

```diff
@@ -105,28 +104,20 @@ impl<'a, DB: WriteResults + Sync> Worker<'a, DB> {
                             &self.name,
                         )?;
                     } else {
-                        self.graph.lock().unwrap().mark_as_completed(id);
-                    }
-
-                    // Unpark all the threads
-                    let mut parked = self.parked_threads.lock().unwrap();
-                    for (_id, thread) in parked.drain() {
-                        thread.unpark();
+                        guard.mark_as_completed(id);
                     }
                 }
                 WalkResult::Blocked => {
-                    // Wait until another thread finished before looking for tasks again
-                    // If the thread spuriously wake up (parking does not guarantee no
-                    // spurious wakeups) it's not a big deal, it will just get parked again
-                    {
-                        let mut parked_threads = self.parked_threads.lock().unwrap();
-                        let current = thread::current();
-                        parked_threads.insert(current.id(), current);
-                    }
-                    thread::park();
+                    guard = self.parked_threads.wait(guard).unwrap();
                 }
                 WalkResult::NotBlocked => unreachable!("NotBlocked leaked from the run"),
-                WalkResult::Finished => break,
+                WalkResult::Finished => {
+                    // A blocked thread may be waiting on the root node, in
+                    // which case this is crucial to avoiding a deadlock.
+                    self.parked_threads.notify_all();
+                    std::mem::drop(guard);
```

Comment:
In case you end up deciding to removing the qualification, this spot is qualified as well.